### PR TITLE
Close FastParquet file even on error

### LINKED ISSUE #46555: BUG: read_parquet leaks file handles on error
### Pandas version checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.
- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

Run this in [development mode](https://docs.python.org/3/library/devmode.html):

```python
from pathlib import Path
import pandas as pd
from pandas.io.parquet import to_parquet, read_parquet

path = Path('/tmp/test.parquet')
to_parquet(pd.DataFrame(dict(a=[1])), path)
with path.open('r+b') as f:
    f.seek(16)
    f.write(b'bogus')

try:
    read_parquet(path)
except Exception:
    pass
```

### Issue Description

The handles aren’t closed in a `except`/`finally` block, so they only get closed when everything works.

When running the above in [development mode](https://docs.python.org/3/library/devmode.html), you’ll see a [`ResourceWarning`](https://docs.python.org/3/library/exceptions.html#ResourceWarning)

### Expected Behavior

all file handles are closed no matter what

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.9.12.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.16.16-zen1-1-zen
Version          : #1 ZEN SMP PREEMPT Mon, 21 Mar 2022 22:59:42 +0000
machine          : x86_64
processor        : 
byteorder        : little
LC_ALL           : None
LANG             : de_DE.UTF-8
LOCALE           : de_DE.UTF-8
pandas           : 1.4.1
numpy            : 1.22.3
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 22.0.4
setuptools       : 57.4.0
Cython           : None
pytest           : 7.1.1
hypothesis       : None
sphinx           : 4.3.2
blosc            : None
feather          : None
xlsxwriter       : 1.4.0
lxml.etree       : 4.6.3
html5lib         : 1.1
pymysql          : None
psycopg2         : 2.8.6
jinja2           : 3.0.3
IPython          : 7.23.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fastparquet      : 0.8.0
fsspec           : 2022.02.0
gcsfs            : None
matplotlib       : 3.4.1
numba            : 0.53.1
numexpr          : 2.7.3
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.8.0
sqlalchemy       : 1.4.23
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : None
xlrd             : 1.2.0
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by flying-sheep**
- [x] closes #46555
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/io/parquet.py [hunk 1]
```diff
@@ -351,11 +351,11 @@ def read(
 
         parquet_file = self.api.ParquetFile(path, **parquet_kwargs)
 
-        result = parquet_file.to_pandas(columns=columns, **kwargs)
-
-        if handles is not None:
-            handles.close()
-        return result
+        try:
+            return parquet_file.to_pandas(columns=columns, **kwargs)
```
Comment (twoertwein) [inline pandas/io/parquet.py:355]: Might as well move `parquet_file = self.api.ParquetFile(path, **parquet_kwargs)` into the try block.

Comment (flying-sheep) [reply pandas/io/parquet.py:354]: Good point, done

Comment (twoertwein) [reply pandas/io/parquet.py:354]: Can you please also add a test. Probably something similar to this test https://github.com/pandas-dev/pandas/blob/3cb70a5f0d1dc386f7806fca75abff6f098b1539/pandas/tests/io/parser/test_unsupported.py#L185

Comment (flying-sheep) [reply pandas/io/parquet.py:354]: Does `ensure_clean` ensure that there’s no file handle to that path?

On Linux, deleting a file doesn’t mean its file handles are closed …

So basically: I’m not sure how to test it.

Comment (twoertwein) [reply pandas/io/parquet.py:354]: `ensure_clean` will try to unlink the file on `__exit__`. This will not detect open files on linux but on windows it will fail. Not the best test but it seems to work :) let me know if you have a better idea!

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -16,7 +16,7 @@ repos:
         pass_filenames: true
         require_serial: false
 -   repo: https://github.com/python/black
-    rev: 22.1.0
+    rev: 22.3.0
```
Comment (jreback) [inline .pre-commit-config.yaml:19]: try to leave out unrelated changes

**Review by jreback — CHANGES_REQUESTED** [review]
going to need a whatsnew note in I/O for 1.5

**Comment by flying-sheep** [comment]
> going to need a whatsnew note in I/O for 1.5

done!

File: pandas/tests/io/test_parquet.py [hunk 3]
```diff
@@ -1155,3 +1155,13 @@ def test_use_nullable_dtypes_not_supported(self, fp):
...
+
+    def test_close_file_handle_on_read_error(self):
+        df = pd.DataFrame({"a": [1, 2]})
+        with tm.ensure_clean() as path:
+            df.to_parquet(path)
+            with open(path, 'r+b') as f:
+                f.seek(16)
+                f.write(b'breakit')
+            with pytest.raises(Exception):  # Not important which exception
+                read_parquet(path, engine="fastparquet")
```
Comment (twoertwein) [inline pandas/tests/io/test_parquet.py:1167]: Two very small comments/suggestions:

- Could use Path(path).write_bytes(b"breakit") (just my personal preference - feel free to ignore)
- Can you explicitly call Path(path).unlink() and add a comment that this would fail on windows if there is still an open file handle (just in case `ensure_clean` changes in the future and catches any exceptions when removing files).

**Review by twoertwein — APPROVED** [review]
Two tiny (non-essential) comments, looks good to me otherwise! Thanks @flying-sheep

File: pandas/tests/io/test_parquet.py (hunk 3 above)
Comment (flying-sheep) [reply pandas/tests/io/test_parquet.py:1163]: great points, done.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @flying-sheep
